File: spark_java/src/main/java/com/spark/sql/study/dataset/RDD2DataSetReflection.java
package com.spark.sql.study.dataset;

import com.spark.sql.study.dataset.vo.Student;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.SparkSession;

import java.util.List;

/**
 *
 * 利用反射的方式
 *
 * @author yangqian
 * @date 2021/5/12
 */
public class RDD2DataSetReflection {

    public static void main(String[] args) {
        SparkConf conf = new SparkConf().setMaster("local").setAppName("RDD2DataSetReflection");
        JavaSparkContext sc = new JavaSparkContext(conf);
        SQLContext sqlContext = SparkSession.builder().config(conf).getOrCreate().sqlContext();
        JavaRDD<String> lines = sc.textFile("/Users/qian/WorkSpaces/own-workspace/2021/spark_code_repository/spark_java/src/main/resources/spark_sql/students.txt");
        JavaRDD<Student> students = lines.map(new Function<String, Student>() {
            private static final long serialVersionUID = 3587019637184285751L;

            @Override
            public Student call(String v1) throws Exception {
                String[] studentInfo = v1.split(",");
                Student student = new Student();
                student.setId(Integer.parseInt(studentInfo[0]));
                student.setName(studentInfo[1]);
                student.setAge(Integer.parseInt(studentInfo[2]));
                return student;
            }
        });

        // 使用反射的方式转化为DataFrame
        // 将 Student.class 传入，其实就是使用反射的方式来创建 DataFrame
        // 因为Student.class本身就是反射的一个应用
        // 底层还是通过 student class 进行反射用于获取field
        // JavaBean必须实现Serializable接口，是可序列化的
        Dataset<Row> studentDF = sqlContext.createDataFrame(students, Student.class);
        // 注册为临时表
        studentDF.registerTempTable("students");
        // 针对students临时表执行sql语句
        Dataset<Row> teenagerDF = sqlContext.sql("select id, name, age from students where age >= 18");
        // 查询后转换为RDD
        JavaRDD<Row> teenagerRDD = teenagerDF.javaRDD();

        // 将RDD中的数据进行映射，成为Student
        JavaRDD<Student> teenagerStudentRDD = teenagerRDD.map(new Function<Row, Student>() {
            @Override
            public Student call(Row row) throws Exception {
                Student stu = new Student();
                stu.setId(row.getInt(0));
                stu.setName(row.getString(1));
                stu.setAge(row.getInt(2));
                return stu;
            }
        });

        // 将数据collect回来，打印出来
        List<Student> studentList = teenagerStudentRDD.collect();
        for (Student student : studentList) {
            System.out.println(student);
        }
    }


}

File: spark_scala/src/main/scala/com/spark/sql/scala/lesson/dataset/RDD2DataSetReflection.scala
package com.spark.sql.scala.lesson.dataset

import com.spark.sql.scala.lesson.dataset.entity.Student
import org.apache.spark.sql.SparkSession
import org.apache.spark.{SparkConf, SparkContext}

/**
 *
 * 要注意区别DataSet 和 DataFrame的区别
 *
 * 利用反射（隐式转换）的方式转为为DataSet
 *
 * @date 2021/5/12
 */
object RDD2DataSetReflection {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("RDD2DataSetReflection").setMaster("local")
    val sc = new SparkContext(conf)
    val sqlContext = SparkSession.builder().config(conf).getOrCreate()

    val studentRDD = sc
      .textFile("/Users/qian/WorkSpaces/own-workspace/2021/spark_code_repository/spark_java/src/main/resources/spark_sql/students.txt")
      .map(line => line.split(","))
      .map(arr => Student(arr(0).toInt, arr(1).toString, arr(2).toInt))

    import sqlContext.implicits._

    val studentDF = sqlContext.createDataset(studentRDD)
    studentDF.printSchema()

    // 与 DataFrame的registerTempTable类似
    studentDF.createOrReplaceTempView("students")

    val teenagerDF = sqlContext.sql("select id, name, age  from students where age <= 18")
    val teenagerRDD = teenagerDF.rdd

    teenagerRDD
      .map(row => Student(row(0).toString.toInt, row(1).toString, row(2).toString.toInt))
      .collect()
      .foreach { stu => println(stu.id + ":" + stu.name + ":" + stu.age) }

    println("=========================================================")

    teenagerRDD.map { row => Student(row.getAs[Int]("id"), row.getAs[String]("name"), row.getAs[Int]("age")) }
      .collect()
      .foreach { stu => println(stu.id + ":" + stu.name + ":" + stu.age) }

    println("=========================================================")

    val studentResultRDD = teenagerRDD.map(row => {
      val map = row.getValuesMap[Any](Array("id", "name", "age"))
      Student(map("id").toString.toInt, map("name").toString, map("age").toString.toInt)
    })

    studentResultRDD.collect().foreach { stu => println(stu.id + ":" + stu.name + ":" + stu.age) }

  }
}
